esp/platform/txsummary.cpp

```diff
@@ -20,22 +20,356 @@
 #include "jutil.hpp"
 #include <algorithm>
 
-using std::find_if;
-using std::for_each;
+#define MATCH_ENTRY [&](const EntryValue& e) {return strieq(e.get()->name.str(), pathPart);}
 
-#define VALIDATE_KEY(k) if (!(k) || !(*k)) return false
-#define MATCH_KEY       [&](const Entry& entry) { return stricmp(entry.key.str(), key) == 0; }
 
-bool operator < (const StringAttr& a, const StringAttr& b)
+inline bool validateKey(const char* k)
 {
-    return (stricmp(a.str(), b.str()) < 0);
+    // Empty or null keys are invalid
+    if(!k || !(*k))
+        return false;
+    // Keys containing an empty path portion
+    // are invalid
+    if(strstr(k, ".."))
+        return false;
+    // Keys beginning or ending in the path
+    // delimiter are considered equivalent to
+    // containing an empty path portion and
+    // therefore are invalid
+    if( ('.' == k[0]) || ('.' == k[strlen(k)-1]))
+        return false;
+    return true;
 }
 
-CTxSummary::CTxSummary(unsigned creationTime)
-: m_creationTime(creationTime ? creationTime : msTick())
+CTxSummary::TxEntryBase::TxEntryBase(const char* _key, const LogLevel _logLevel, const unsigned int _group, const char* _suffix): logLevel(_logLevel), group(_group), suffix(_suffix)
+{
+    // parse dot-delimited key
+    // 'name' is set to the rightmost element of _key
+    // 'fullname' is the entire _key
+
+    if(isEmptyString(_key))
+        name.set(_key);
+    else
+    {
+        const char* finger = strrchr(_key, '.');
```

Comment:
What's the expect prevalence for keys without periods? Setting two copies of the same string is really wasteful. Setting a copy of a full string plus a copy of the tail end of the same string is just wasteful. My first reaction was to suggesting using String to avoid two copies of the same text, but only one copy of the full text is actually needed. Make `name` a `const char*` that points into the `fullname` buffer.

Maybe this is a premature optimization, but eliminating unnecessary memory allocations and a reduction in memory consumption might make it worthwhile.

Reply:
That seems reasonable, I'll make the change.

Reply:
OK that sounds good, and seems fairly straightforward.